### ISSUE TITLE
inference failed with a custom `example_inference.py`

### ISSUE THREAD

**Issue by yo3nglau** [DeweiHu/AdaptDiff#3]
Hello Dewei,

thanks for your great work.

I noticed that the `example_inference.py` file is still in a phase of **TODO**, so I implemented a simple code to show the generation results given binary masks. I referred to the codes of Lines 93-105 in `conditional_diffusion_trainer.py` and loaded the released checkpoint for the `OCTA-500` dataset. I simply used the ground-truth masks of three images in the `OCTA-500` dataset as the `condition` and the images as the `template` to match the histogram. Confusingly, **the generation results do not make sense**. I attached the implementation below and look forward to your response.

![0](https://github.com/user-attachments/assets/6d20b14c-3c35-4858-abe6-0e2b22dde2f0)
![1](https://github.com/user-attachments/assets/41f3e84d-1740-4b5d-a6c9-c77dce42e0af)
![2](https://github.com/user-attachments/assets/9d25b6fc-6c26-419d-8074-6bc2baa59d0a)

    import os
    import torch
    import numpy as np
    from pathlib import Path
    from PIL import Image
    
    import diffusion_solver
    import models
    import utils
    
    
    device = torch.device("cuda")
    
    # load data
    source_dir = Path("AdaptDiff/exp/inference/")
    images_t_dir = source_dir / "target_images"
    masks_s_dir = source_dir / "source_masks"
    
    inference_data = []
    for masks_s_name in os.listdir(masks_s_dir):
        images_s = Image.open(images_t_dir / masks_s_name).convert('L').resize((384, 384))
        masks_s = Image.open(masks_s_dir / masks_s_name).convert('L').resize((384, 384))
        inference_data.append((images_s, masks_s))
    
    # load model
    model = models.condition_Unet().to(device)
    model.load_state_dict(torch.load(f"AdaptDiff/diffusion.octa500.pt"))
    
    # diffusion configuration
    beta_start = 0.0001
    beta_end = 0.02
    T = 300
    betas = diffusion_solver.get_beta_schedule(beta_schedule="linear",
                                               beta_start=beta_start,
                                               beta_end=beta_end,
                                               num_diffusion_timesteps=T)
    
    sampler = diffusion_solver.DiffusionSampler(betas,
                                                device=device,
                                                mode='conditional')
    
    syn_pairs = []
    
    for idx, (im_t, x_condition) in enumerate(inference_data):
        x_t = torch.randn((1, 1, 384, 384))
        x_condition = torch.tensor(np.array(x_condition), dtype=torch.float)
        x_condition = x_condition[None, None, :, :]
        syn_im_t, mask_s = sampler.get_conditonal_diffusion_result(x_t, x_condition, model)
    
        syn_im_t = utils.ImageRescale(syn_im_t, [0, 1])
        template = utils.ImageRescale(np.array(im_t), [0, 1])
        hm_syn_im_t = utils.hist_match(syn_im_t, template)
    
        hm_syn_im_t = np.uint8(utils.ImageRescale(hm_syn_im_t, [0, 255]))
        mask_s = np.uint8(utils.ImageRescale(mask_s, [0, 255]))
    
        outputs_dir = source_dir / "synthesized_pairs"
        os.makedirs(outputs_dir, exist_ok=True)
    
        utils.image_saver(np.hstack((hm_syn_im_t, mask_s)), outputs_dir, idx)

**Comment by DeweiHu** [comment]
Hi there,

I am sorry for the delay; I will finish the reference code later this week. A further notice will be sent.

Thanks!

Best,
Dewei

From: yo3nglau ***@***.***>
Date: Monday, October 28, 2024 at 5:39 AM
To: DeweiHu/AdaptDiff ***@***.***>
Cc: Subscribed ***@***.***>
Subject: [DeweiHu/AdaptDiff] inference failed with a custom `example_inference.py` (Issue #3)

Hello Dewei,

thanks for your great work.

I noticed that the example_inference.py file is still in a phase of TODO, so I implemented a simple code to show the generation results given binary masks. I referred to the codes of Lines 93-105 in conditional_diffusion_trainer.py and loaded the released checkpoint for the OCTA-500 dataset. I simply used the ground-truth masks of three images in the OCTA-500 dataset as the condition and the images as the template to match the histogram. Confusingly, the generation results do not make sense. I attached the implementation below and look forward to your response.

0.png (view on web)<https://github.com/user-attachments/assets/6d20b14c-3c35-4858-abe6-0e2b22dde2f0>
1.png (view on web)<https://github.com/user-attachments/assets/41f3e84d-1740-4b5d-a6c9-c77dce42e0af>
2.png (view on web)<https://github.com/user-attachments/assets/9d25b6fc-6c26-419d-8074-6bc2baa59d0a>

import os

import torch

import numpy as np

from pathlib import Path

from PIL import Image



import diffusion_solver

import models

import utils





device = torch.device("cuda")



# load data

source_dir = Path("AdaptDiff/exp/inference/")

images_t_dir = source_dir / "target_images"

masks_s_dir = source_dir / "source_masks"



inference_data = []

for masks_s_name in os.listdir(masks_s_dir):

    images_s = Image.open(images_t_dir / masks_s_name).convert('L').resize((384, 384))

    masks_s = Image.open(masks_s_dir / masks_s_name).convert('L').resize((384, 384))

    inference_data.append((images_s, masks_s))



# load model

model = models.condition_Unet().to(device)

model.load_state_dict(torch.load(f"AdaptDiff/diffusion.octa500.pt"))



# diffusion configuration

beta_start = 0.0001

beta_end = 0.02

T = 300

betas = diffusion_solver.get_beta_schedule(beta_schedule="linear",

                                           beta_start=beta_start,

                                           beta_end=beta_end,

                                           num_diffusion_timesteps=T)



sampler = diffusion_solver.DiffusionSampler(betas,

                                            device=device,

                                            mode='conditional')



syn_pairs = []



for idx, (im_t, x_condition) in enumerate(inference_data):

    x_t = torch.randn((1, 1, 384, 384))

    x_condition = torch.tensor(np.array(x_condition), dtype=torch.float)

    x_condition = x_condition[None, None, :, :]

    syn_im_t, mask_s = sampler.get_conditonal_diffusion_result(x_t, x_condition, model)



    syn_im_t = utils.ImageRescale(syn_im_t, [0, 1])

    template = utils.ImageRescale(np.array(im_t), [0, 1])

    hm_syn_im_t = utils.hist_match(syn_im_t, template)



    hm_syn_im_t = np.uint8(utils.ImageRescale(hm_syn_im_t, [0, 255]))

    mask_s = np.uint8(utils.ImageRescale(mask_s, [0, 255]))



    outputs_dir = source_dir / "synthesized_pairs"

    os.makedirs(outputs_dir, exist_ok=True)



    utils.image_saver(np.hstack((hm_syn_im_t, mask_s)), outputs_dir, idx)

**Comment by yo3nglau** [comment]
Thank you for your quick response! I will stay tuned.

**Comment by DeweiHu** [comment]
Hello, just a quick headsup, I have upload the adapt_inference.py and run_inference.sh. You can try to inference the model with your own binary mask in shape (256, 256) or (384, 384). Or you can try the mask provided (binary_mask.png). Please let me know if you have any other questions.

**Comment by yo3nglau** [comment]
Thank you for your guidance! I will try them asap.

**Comment by yo3nglau** [comment]
I made a successful inference given the newly released codes. This issue is closed with thanks now. I am going to re-train the model by myself. I would open new issues if I have any further questions. Thanks again for your effective guidance.

**Comment by DeweiHu** [comment]
Sounds good, feel free to reach out anytime.

Get Outlook for iOS<https://aka.ms/o0ukef>
________________________________
From: yo3nglau ***@***.***>
Sent: Monday, November 4, 2024 9:45:39 PM
To: DeweiHu/AdaptDiff ***@***.***>
Cc: Hu, Dewei, Ph.D., M.E. ***@***.***>; Comment ***@***.***>
Subject: [EXTERNAL] Re: [DeweiHu/AdaptDiff] inference failed with a custom `example_inference.py` (Issue #3)


I made a successful inference given the newly released codes. This issue is closed with thanks now. I am going to re-train the model by myself. I would open new issues if I have any further questions. Thanks again for your effective guidance.
